overlord/snapstate/handlers.go

```diff
@@ -327,9 +327,17 @@ func (m *SnapManager) installPrereqs(t *state.Task, base string, prereq []string
 		if err != nil {
 			return prereqError("prerequisite", prereqName, err)
 		}
-		if ts == nil {
+		if ts == nil || len(ts.Tasks()) == 0 {
 			continue
 		}
+
+		// make the dependencies of the "prerequisite" task wait for the install tasks. Otherwise, the main snap will be
+		// installed in parallel with its prerequisites, resulting in a race condition. Waiting for the last task is
+		// enough since tasks wait for the previous one to be done before running
+		for _, dep := range t.HaltTasks() {
+			lastPrereqTask := ts.Tasks()[len(ts.Tasks())-1]
```

Comment:
Sorry for realizing it only now, but this line could actually be moved outside of the loop. This will reduce the calls to `ts.Tasks()`. Then, they could be reduced even more (we would anyway call it three times: two to get the length, and once to extract the last element) by calling it only once and cache its value, but I'm no golang expert so I'm not sure if this is worth it or if the golang compiler is smart enough to optimize away the extra calls.

Reply:
Thanks!